### DISCUSSION TITLE
Proporsal: Static container type for static abstract

### DISCUSSION THREAD

**Discussion by Vixenka** [dotnet/csharplang#6973] [Language Ideas]
# Introduction
Together with C# 11 and .NET 7, [static abstract](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-11.0/static-abstracts-in-interfaces) language function appeared. Which allow us to create static polymorphism on interfaces.

# Problem
Currently, when we call static abstract things, we do it on a generic type. But I miss the existence of a container to store these types in static form for later invocation.

Is impossible to do something like this without instantiating the type:
```cs
interface IFoo
{
    static abstract void DoSomething();
}

class Bar : IFoo
{
    public static void DoSomething()
    {
        System.Console.WriteLine("wow");
    }
}

static class Invoker
{
    private static readonly List<Static<IFoo>> foos = new();
    
    public static void Add<T>() where T : IFoo 
    {
        foos.Add(Static<IFoo>.Create<T>()); // There is currently no way to store a static T type without instantiating it.
    }
    
    public static void Invoke()
    {
        foreach (var foo in foos)
           foo.DoSomething();
    }
}
```

# Alternatives
Currently this is possible to perform a similar type by source generators and reflection.

# Solution
My suggestion for solving this problem is adding type `Static<T>` same as in the code shown in the problem.

# Implementation problems
Creating a new special type that cannot be created directly with the language syntax. (Unless using source generators or reflection, but this is slow to call and does not necessarily create compile errors.)

**Comment by HaloFour** [comment]
You could store the static methods as delegates:

```cs
using System;
using System.Collections.Generic;

public interface IFoo {
    static abstract void DoSomething();
}


public class Foo : IFoo {
    public static void DoSomething() => Console.WriteLine("Foo");
}

public class Bar : IFoo {
    public static void DoSomething() => Console.WriteLine("Bar");
}


public class Container {
    private static List<Action> list = new List<Action>();
    
    public static void Add<T>() where T : IFoo {
        list.Add(T.DoSomething);
    }
    
    public static void InvokeAll() {
        foreach (var action in list) {
            action();
        }
    }
}

static class Program {
    static void Main() {
        Container.Add<Foo>();
        Container.Add<Bar>();
        
        Container.InvokeAll();
    }
}
```

**Reply by Vixenka** [reply]
You right, I mentioned this in alternatives. But when we want to use several such methods in different types, we must write or use souce generator, where Static<T> would be give this automatically.
